module/zfs/dbuf.c

```diff
@@ -1841,16 +1888,24 @@ dbuf_read(dmu_buf_impl_t *db, zio_t *pio, uint32_t flags)
 		}
 		mutex_exit(&db->db_mtx);
 	} else {
+		blkptr_t *bp = NULL;
 		ASSERT(db->db_state == DB_UNCACHED ||
 		    db->db_state == DB_NOFILL);
 		db_lock_type_t dblt = dmu_buf_lock_parent(db, RW_READER, FTAG);
+
+		/*
+		 * If a Direct I/O write has occurred we will use the updated
+		 * block pointer.
+		 */
+		bp = dmu_buf_get_bp_from_dbuf(db);
```

Comment:
So I have been looking this. I was trying to think of a nice way to unify the code, but I honestly don't think it is an issue the way it stands. So in the `O_DIRECT` case, there is never a time that a read could occur while the `O_DIRECT` write is happening. They only happen across entire blocks to avoid the RWM and the rangelock is held across the block. So there is now way in `dbuf_read()` or `dbuf_read_impl()`, that an `O_DRIECT` write could be tied to the `db_state` of `DB_NOFILL`.

The reason I pull the block pointer early to check for the `O_DIRECT` write in `dbuf_read()` is because of the check:
```
                if (pio == NULL && (db->db_state == DB_NOFILL ||
                    (bp != NULL && !BP_IS_HOLE(bp)))) {
                        spa_t *spa = dn->dn_objset->os_spa;
                        pio = zio_root(spa, NULL, NULL, ZIO_FLAG_CANFAIL);
                        need_wait = B_TRUE;
                }    

```
The old check just relied on the `db->db_blkptr` to determine if the `need_wait` was necessary. In the block cloning case, the second check short circuits in the if `db->db_state == DB_NOFILL` so the overridden BP is not needed at that time. I am open for suggestions here though as I have only looked at this for a bit. I am good with the way the code is but it definitely is not unifying the two in this code path as it stands now.

Reply:
Both `dmu_buf_get_bp_from_dbuf()` and a chunk of `dbuf_read_impl()` read block pointer from dr_overridden_by of dirty record.  Just one handles only direct I/O case, while another -- only block cloning.  To me this looks like a duplication.  If you want to get the data earlier -- that is fine.  Speaking about `db->db_state == DB_NOFILL` check, I think you are actually missing it in your code.  The fact that the last dirty record was an override does not mean that you need to get block pointer from it, since buffer may have already be read after that and state could be changed from `DB_NOFILL`.

PS: And also `dbuf_get_dirty_direct()` looks like obfuscation to me.  It has nothing specific to direct I/O, so I would inline it to make obvious what it is doing.

Reply:
So, I updated this in the latest commit to unify both code paths. Let me know if this looks good to you before I resolve this comment. Also, I got rid of `dbuf_get_dirty_direct()`.

Reply:
I'll look on it more, but I wonder why dbufs in case of Direct I/O use DB_UNCACHED state, while in case of block cloning DB_NOFILL.  I'd expect them to be the same.  In dmu_buf_get_bp_from_dbuf() if you pull `dr != NULL` check one level up, the cases would also get a bit closer.

Reply:
Yep, made sense to pull up the `dr` check another level. I just squashed this into the latest commit.

Reply:
I meant it just as a first step to further unify the two cases, since in both cases you return `&dr->dt.dl.dr_overridden_by`.

Just cosmetically I would invert the `if (dr)` condition to reduce indentation.

Reply:
So I did a deep dive on this to further understand the block cloning reasoning, and the latest commit gets closer to unifying the two. So the big different between the two is that block cloning relies on always calling `dbuf_undiryt()` so that `brt_pending_remove()` is called. In the Direct I/O case, it relies on `dbuf_unoverride()` being called for the cleanup on space accounting that takes place in with `zio_free()`. That is the reason Direct I/O transitions to `DB_UNCACHED` after it has completed. This allows for `dbuf_unoverride()` to be called even in the case of `dbuf_undirty()`. The other part of this is that Direct I/O does not need to call `dbuf_undirty()`, which is pretty heavy weight, when compared to just needing to call `dbuf_redirty()` though the dbuf interfaces, which winds up calling `dbuf_unoverrideI(). I think block cloning relies on calling `dbuf_undirty()` to always  make the call to `brt_pending_remove()` for any cleanup that has to happen for that case.

With that all that being said, I think the latest PR handles the two and unifies them much more. I also have toyed around with using an `uint8_t` as a flag in the dirty record leaf to differentiate between block cloning and Direct I/O instead of two `boolean_t`'s in the union. Originally @behlendorf and I avoided making the dirty record struct larger than it was previously. I have the patch ready if we want that. I also created another version of the unification where Direct I/O stays in the `DB_NOFILL` state:
https://github.com/bwatkinson/zfs/commit/d5e60f51005f3d075b27ae6abd7c879b5d891e3c
However, it I don't personally think it is any better than what I have pushed to this PR. It also adds more logic and really winds up being six of one, half a dozen of the other.